File: klinara-ios/klinara-ios/Features/Packages/BindPackageSheet.swift
import SwiftUI

/// Randevu kalemini bir paket hakkına bağlar.
///
/// Randevu **henüz tamamlanmadıysa** yalnız bağlar; seans, durum `completed`
/// olduğunda aynı transaction'da düşer. Tamamlanmışsa bağlar VE düşer —
/// sunucu ikisini tek uçta yapıyor, bu ekran farkı kullanıcıya yazıyor.
struct BindPackageSheet: View {

    let session: AppSession
    let appointment: Appointment
    let line: AppointmentServiceLine
    var onBound: () -> Void

    @Environment(\.dismiss) private var dismiss

    @State private var entitlements: LoadState<[PackageEntitlement]> = .loading
    @State private var selectedItemId: String?
    @State private var error: APIError?
    @State private var isSaving = false
    @State private var idempotencyKey = UUID().uuidString

    private var clock: BranchClock { session.clock }

    var body: some View {
        NavigationStack {
            KlinaraScreen(
                state: entitlements,
                emptyCheck: { $0.isEmpty },
                emptyTitle: "Kullanılabilir hak yok",
                emptyMessage: "Müşterinin bu hizmet için aktif ve süresi dolmamış paketi bulunmuyor.",
                emptyIcon: "shippingbox",
                onRetry: load
            ) { options in
                if let error, !error.isFieldScoped {
                    ErrorBanner(error: error)
                }

                KlinaraCard(title: "Paket hakları", footnote: footnote) {
                    ForEach(Array(options.enumerated()), id: \.element.id) { index, item in
                        if index > 0 { KlinaraDivider() }
                        Button {
                            selectedItemId = item.customerPackageItemId
                        } label: {
                            KlinaraRow(label: item.packageName, detail: detail(item)) {
                                if selectedItemId == item.customerPackageItemId {
                                    Image(systemName: "checkmark")
                                        .font(.system(size: 14, weight: .semibold))
                                        .foregroundStyle(KlinaraColor.sageDeep)
                                }
                            }
                        }
                        .buttonStyle(.plain)
                    }
                }

                KlinaraButton(
                    title: appointment.status == .completed ? "Bağla ve düş" : "Pakete bağla",
                    kind: .primary,
                    icon: "shippingbox",
                    isLoading: isSaving,
                    isEnabled: selectedItemId != nil && !isSaving
                ) {
                    Task { await bind() }
                }
            }
            .navigationTitle("Pakete bağla")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarLeading) {
                    Button("Vazgeç") { dismiss() }
                        .klinaraText(.bodyM)
                        .foregroundStyle(KlinaraColor.charcoalMuted)
                }
            }
            .task { await load() }
        }
        .tint(KlinaraColor.sage)
    }

    private var footnote: String {
        appointment.status == .completed
            ? "Randevu tamamlandığı için seans HEMEN düşer."
            : "Seans, randevu tamamlandığında düşer."
    }

    private func detail(_ item: PackageEntitlement) -> String {
        var parts = ["\(item.remainingSessions) seans kaldı"]
        if let expiresAt = item.expiresAt {
            parts.append("\(clock.formatDate(expiresAt))'de doluyor")
        }
        return parts.joined(separator: " · ")
    }

    private func load() async {
        entitlements = .loading
        do {
            // Yalnız bu hizmetin hakları: başka bir hizmetin kalemine bağlamak
            // sunucuda zaten reddediliyor, listede göstermek boşuna umut.
            let options = try await session.services.packages.entitlements(
                customerId: appointment.customerId,
                serviceId: line.serviceId,
                branchId: appointment.branchId
            )
            entitlements = .loaded(options)
        } catch {
            entitlements = .failed(error as? APIError ?? .network)
        }
    }

    private func bind() async {
        guard let itemId = selectedItemId else { return }
        error = nil
        isSaving = true
        defer { isSaving = false }
        do {
            _ = try await session.services.packages.consume(
                appointmentId: appointment.id,
                ConsumePackageInput(lines: [
                    ConsumePackageLineInput(
                        appointmentServiceId: line.id,
                        customerPackageItemId: itemId
                    )
                ]),
                idempotencyKey: idempotencyKey
            )
            onBound()
            dismiss()
        } catch {
            self.error = error as? APIError ?? .network
        }
    }
}

File: klinara-android/app/src/main/java/com/klinara/android/features/packages/BindPackageSheet.kt
package com.klinara.android.features.packages

import androidx.compose.foundation.layout.Box
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ShoppingCart
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import com.klinara.android.designsystem.KlinaraTheme
import com.klinara.android.designsystem.KlinaraType
import com.klinara.android.designsystem.components.AuthLoadingOverlay
import com.klinara.android.designsystem.components.EmptyStateView
import com.klinara.android.designsystem.components.ErrorBanner
import com.klinara.android.designsystem.components.KlinaraButton
import com.klinara.android.designsystem.components.KlinaraCard
import com.klinara.android.designsystem.components.KlinaraDivider
import com.klinara.android.designsystem.components.KlinaraScreen
import com.klinara.android.designsystem.components.KlinaraSelectableRow
import com.klinara.android.designsystem.components.KlinaraSkeleton
import com.klinara.android.designsystem.components.KlinaraSkeletonStyle
import com.klinara.android.services.formatting.BranchClock
import com.klinara.android.services.networking.Loadable
import com.klinara.android.services.packages.PackageEntitlement

/**
 * "Pakete bağla" — randevu detayındaki hizmet satırından açılan hedef.
 *
 * Randevu **henüz tamamlanmadıysa** yalnız bağlar; seans, durum `completed` olduğunda
 * aynı transaction'da düşer. Tamamlanmışsa bağlar VE düşer. Sunucu ikisini tek uçta
 * yapıyor; bu ekran farkı kullanıcıya yazıyor.
 */
@Composable
fun BindPackageSheet(
    state: BindPackageUiState,
    serviceName: String?,
    clock: BranchClock,
    onSelect: (String) -> Unit,
    onBind: () -> Unit,
    onRetry: () -> Unit,
    onDismissError: () -> Unit,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Box {
        KlinaraScreen(title = "Pakete bağla", modifier = modifier, onBack = onBack) {
            state.error?.let { ErrorBanner(message = it, retryLabel = "Kapat", onRetry = onDismissError) }

            when (val options = state.entitlements) {
                Loadable.Loading ->
                    KlinaraSkeleton(style = KlinaraSkeletonStyle.rowsShort)
                is Loadable.Failed ->
                    ErrorBanner(message = options.message, onRetry = if (options.isRetryable) onRetry else null)
                is Loadable.Loaded ->
                    if (options.value.isEmpty()) {
                        EmptyStateView(
                            title = "Kullanılabilir hak yok",
                            message =
                                "Müşterinin ${serviceName ?: "bu hizmet"} için aktif ve süresi dolmamış " +
                                    "paket hakkı bulunmuyor.",
                            icon = Icons.Filled.ShoppingCart,
                        )
                    } else {
                        KlinaraCard(
                            title = serviceName?.let { "$it için haklar" } ?: "Paket hakları",
                            footnote =
                                if (state.consumesNow) {
                                    "Randevu tamamlandığı için seans HEMEN düşer."
                                } else {
                                    "Seans, randevu tamamlandığında düşer."
                                },
                        ) {
                            options.value.forEachIndexed { index, entitlement ->
                                if (index > 0) KlinaraDivider()
                                KlinaraSelectableRow(
                                    title = entitlement.packageName,
                                    detail = detail(entitlement, clock),
                                    isSelected = entitlement.customerPackageItemId == state.selectedItemId,
                                    onClick = { onSelect(entitlement.customerPackageItemId) },
                                )
                            }
                        }
                        KlinaraButton(
                            title = if (state.consumesNow) "Bağla ve düş" else "Pakete bağla",
                            onClick = onBind,
                            isLoading = state.isSaving,
                            enabled = state.canBind,
                        )
                    }
            }
        }

        if (state.isSaving) AuthLoadingOverlay(message = "Bağlanıyor…")
    }
}

private fun detail(
    entitlement: PackageEntitlement,
    clock: BranchClock,
): String =
    listOfNotNull(
        "${entitlement.remainingSessions} seans kaldı",
        entitlement.expiresAt?.let { "${clock.formatDate(it)} tarihinde doluyor" },
    ).joinToString(" · ")
